AmplifyPlugins/API/AWSAPICategoryPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift

```diff
@@ -34,15 +60,82 @@ public struct AWSAPICategoryPluginConfiguration {
         let endpoints = try AWSAPICategoryPluginConfiguration.endpointsFromConfig(config: config,
                                                                                   apiAuthProviderFactory: apiAuthProviderFactory,
                                                                                   authService: authService)
-        self.init(endpoints: endpoints)
+        let interceptors = try AWSAPICategoryPluginConfiguration.interceptorsForEndpoints(endpoints,
+                                                     apiAuthProviderFactory: apiAuthProviderFactory,
+                                                     authService: authService)
+
+        self.init(endpoints: endpoints,
+                  interceptors: interceptors,
+                  apiAuthProviderFactory: apiAuthProviderFactory,
+                  authService: authService)
+
+    }
+
+    /// Registers an interceptor for the provided API endpoint
+    /// - Parameter interceptor: operation interceptor used to decorate API requests
+    /// - Parameter toEndpoint: API endpoint name
+    mutating func addInterceptor(_ interceptor: URLRequestInterceptor,
+                                 toEndpoint apiName: APIEndpointName) {
+        interceptors[apiName]?.addInterceptor(interceptor)
+    }
+
+    /// Returns all the interceptors registered for `apiName` API endpoint
+    /// - Parameter apiName: API endpoint name
+    /// - Returns: request interceptors
+    internal func interceptorsForEndpoint(named apiName: APIEndpointName?) -> [URLRequestInterceptor] {
+        guard let apiName = apiName, let interceptorsConfig = interceptors[apiName] else {
+            return []
+        }
+        return interceptorsConfig.interceptors
+    }
+
+    /// Returns or create interceptors for the provided endpointConfig
+    /// - Parameters:
+    ///   - endpointConfig: endpoint configuration
+    ///   - authType: if provided override the registered auth interceptor
```

Comment:
will there ever be code that dynamically passes in a `nil` authType? If no, callers making the call like `try pluginConfig.interceptorsForEndpoint(withConfig: endpointConfig)` in AWSGraphQLOperation should just call the one without the default parameter: `try pluginConfig.interceptorsForEndpoint(named: endpointConfig.name)`, and then make this API's authType a required parameter. That would simplify the logic to throw when `apiAuthProviderFactory` cannot be unwrapped.

`interceptorsForEndpoint(named apiName: APIEndpointName)` returns all the interceptors for the apiEndpoint

`interceptorsForEndpoint(withConfig endpointConfig: EndpointConfig,
                                          authType: AWSAuthorizationType)`  returns the interceptor for the specified auth type